src/ai/backend/common/metrics/metric.py

```diff
@@ -489,3 +494,29 @@ def instance(cls):
     def to_prometheus(self) -> str:
         self.system.observe()
         return generate_latest().decode("utf-8")
+
+
+class StageObserver:
+    _instance: Optional[Self] = None
+
+    _stage_count: Counter
+
+    def __init__(self) -> None:
+        self._stage_count = Counter(
+            name="backendai_stage_count",
+            documentation="Count stage occurrences",
+            labelnames=["stage_name", "metric_domain", "metric_domain_id"],
+        )
+
+    @classmethod
+    def instance(cls) -> Self:
+        if cls._instance is None:
+            cls._instance = cls()
+        return cls._instance
+
+    def observe_stage(
+        self, *, stage_name: str, metric_domain: StageMetricDomain, metric_domain_id: str
+    ) -> None:
+        self._stage_count.labels(
+            stage_name=stage_name, metric_domain=metric_domain, metric_domain_id=metric_domain_id
```

Comment:
It doesn't seem like the metric_domain is really necessary. It seems like it would be good to only have the stage_name and the upper layer information where the stage is executed.